### PR TITLE
[SPARK-39805][SS] Deprecate Trigger.Once and Promote Trigger.AvailableNow

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to deprecate Trigger.Once and suggest Trigger.AvailableNow as a replacement.

This PR also tries to replace Trigger.Once to Trigger.AvailableNow in the test code as well, except the cases Trigger.Once is used intentionally.

### Why are the changes needed?

Trigger.Once() exposes various issues, including:

1) weak guarantee of the contract

This is the javadoc content of `Trigger.Once`:

> A trigger that processes all available data in a single batch then terminates the query.

Spark does not respect the contract when there is "uncommitted" batch in the previous run. It really works as the name represents, "just run a single batch", hence if there is "uncommitted" batch, Spark will execute the "uncommitted" batch and terminate without processing new data.

2) scalable issue on batch

This is the main rationalization we introduced Trigger.AvailableNow.

3) huge output latency for stateful operator due to the lack of no-data batch

Since Trigger.Once executes the single batch and terminates, the processing for watermark advancement is deferred to the next execution of the query, which tends to be multiple hours or even day(s).

### Does this PR introduce _any_ user-facing change?

Yes, end users will start to see the deprecation message when they use Trigger.Once. The deprecation message guides the end users to migrate to Trigger.Available, with the rationalization on migration.

### How was this patch tested?

Existing UTs

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala:272]: Here the change is actually based on the good thing being provided by Trigger.AvailableNow - Trigger.AvailableNow deals with no-data batch as well, hence the output is being emitted earlier.

**Comment by HeartSaVioR** [comment]
Github Action only fails with [Run TPC-DS queries with SF=1](https://github.com/HeartSaVioR/spark/runs/7401791317?check_suite_focus=true#logs)
We can retrigger once the PR gets reviewed and getting approval.

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @viirya @xuanyuanking @bozhang2820

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (LuciferYang) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:341]: hmm... should we collect them in a new `DeprecatedXXSuite`, which may be easier to notice than comments

File: docs/structured-streaming-programming-guide.md
Comment (viirya) [inline docs/structured-streaming-programming-guide.md:2832]: Hmm, but we still have "One-time micro-batch" in the doc. Do you want to remove it or mark it as deprecated too?

**Review by viirya — COMMENTED** [review]
I think we also need to update the programming guide doc.

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala
Comment (HeartSaVioR) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:341]: Probably a good idea for SQL core module but it does not work with connectors. I'm thinking of another mode of Trigger.AvailableNow to behave like Trigger.Once for processing available data (deal with one batch ignoring source options), and then these usages can be replaced with new mode.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:2832]: Yeah not sure what would be a good way to do. While we can do aggressively via removing all occurrences about Trigger.Once, probably marking as deprecated without removing the content might make more sense as we still have users leveraging Trigger.Once.

Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: I just dealt with mentioning deprecation, and more information in Trigger.AvailableNow to produce motivation to migrate.

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
`ParquetIOSuite.SPARK-7837 Do not close output writer twice when commitTask() fails`

Hit this again. Will retrigger.

**Comment by HeartSaVioR** [comment]
It only failed in TPC-DS (again) which does not run any streaming query AFAIK. I'm ignoring it.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
